src/transactions/TransactionUtils.cpp

```diff
@@ -1172,6 +1174,321 @@ removeOffersByAccountAndAsset(AbstractLedgerTxn& ltx, AccountID const& account,
+                                               trustAcc);
+
+            // using poolTL() after this will throw an exception
+            poolShareTrustLine.erase();
+        }
+
+        auto redeemIntoClaimableBalance =
+            [&ltxInner, &txSourceID, txSeqNum, opIndex, &poolID, &accountID,
+             &cbSponsoringAccID](Asset const& assetInPool,
+                                 int64_t amount) -> bool {
+            // if the claimant is the issuer, then we don't create the claimable
+            // balance
+            if (isIssuer(accountID, assetInPool))
+            {
+                return true;
+            }
+
+            // create a claimable balance for the asset in the pool
+            LedgerEntry claimableBalance;
+            claimableBalance.data.type(CLAIMABLE_BALANCE);
+
+            auto& claimableBalanceEntry =
+                claimableBalance.data.claimableBalance();
+
+            claimableBalanceEntry.balanceID =
+                getRevokeID(txSourceID, txSeqNum, opIndex, poolID, assetInPool);
+            claimableBalanceEntry.amount = amount;
+            claimableBalanceEntry.asset = assetInPool;
+            claimableBalanceEntry.claimants = {
+                makeUnconditionalClaimant(accountID)};
+
+            // if this asset isn't native
+            // 1. set clawback if it's set on the trustline
+            // 2. decrement liquidityPoolUseCount on the asset trustline
+            if (assetInPool.type() != ASSET_TYPE_NATIVE)
+            {
+                // trustline must exist
+                auto assetTrustLine =
+                    loadTrustLine(ltxInner, accountID, assetInPool);
+                if (assetTrustLine.isClawbackEnabled())
+                {
+                    setClaimableBalanceClawbackEnabled(claimableBalanceEntry);
+                }
+            }
+
+            // The account that is currently sponsoring the soon to be
+            // deleted pool share trustline will be used for the claimable
+            // balances. If that account is currently in a sponsorship
+            // sandwich, that sponsoring account will instead sponsor the
+            // claimable balance.
+            auto cbSponsoringLtxAcc =
+                stellar::loadAccount(ltxInner, cbSponsoringAccID);
+            switch (createEntryWithPossibleSponsorship(
+                ltxInner, ltxInner.loadHeader(), claimableBalance,
+                cbSponsoringLtxAcc))
+            {
+            case SponsorshipResult::SUCCESS:
+                break;
+            case SponsorshipResult::LOW_RESERVE:
+                // This is possible only if the account that paid the
+                // reserve for the pool share trustline above is in a
+                // sponsorship sandwich
+                return false;
+            case SponsorshipResult::TOO_MANY_SPONSORING:
+                // This is impossible because numSponsoring should always
```

Comment:
(CAP question I guess)

Is that true? If there is a sponsorship sandwich, we're releasing an entry from `cbSponsoringAccID` but adding one to the sponsor

Reply:
You're right - If the sponsoring account has `numSponsoring + numSubentries == UINT32_MAX` (this limit is in the CAP but not master yet), then this will throw. We should instead set `opTOO_MANY_SPONSORING` like we do everywhere else.

I'll update the CAP as well. The issuer just needs to submit the operation outside of a sponsorship sandwich for this to succeed.

Reply:
https://github.com/stellar/stellar-protocol/pull/1044